server/src/main/java/io/crate/execution/engine/aggregation/impl/average/numeric/NumericAverageAggregation.java

```diff
@@ -0,0 +1,339 @@
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.function.Function;
+
+import static io.crate.execution.engine.aggregation.impl.average.AverageAggregation.NAMES;
+import static io.crate.execution.engine.aggregation.impl.average.numeric.NumericAverageStateType.INIT_SIZE;
+
+public class NumericAverageAggregation extends AggregationFunction<NumericAverageState, BigDecimal> {
+
+    static {
+        DataTypes.register(NumericAverageStateType.ID, in -> NumericAverageStateType.INSTANCE);
+    }
+
+    public static void register(AggregationImplModule mod) {
+        for (var functionName : NAMES) {
+            mod.register(
+                Signature.aggregate(
+                    functionName,
+                    DataTypes.NUMERIC.getTypeSignature(),
+                    DataTypes.NUMERIC.getTypeSignature()),
+                NumericAverageAggregation::new
+            );
+        }
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+    private final DataType<BigDecimal> returnType;
+
+    @VisibleForTesting
+    private NumericAverageAggregation(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+
+        // We want to preserve the scale and precision from the
+        // numeric argument type for the return type. So we use
+        // the incoming numeric type as return type instead of
+        // the return type from the signature `avg(count::numeric(16, 2))`
+        // should return the type `numeric(16, 2)` not `numeric`
+        var argumentType = boundSignature.getArgumentDataTypes().get(0);
+        assert argumentType.id() == DataTypes.NUMERIC.id();
+        //noinspection unchecked
+        this.returnType = (DataType<BigDecimal>) argumentType;
+    }
+
+    @Nullable
+    @Override
+    public NumericAverageState newState(RamAccounting ramAccounting,
+                               Version indexVersionCreated,
+                               Version minNodeInCluster,
+                               MemoryManager memoryManager) {
+        ramAccounting.addBytes(INIT_SIZE);
+        return new NumericAverageState(new BigDecimalValueWrapper(BigDecimal.ZERO), 0L);
+    }
+
+    @Override
+    public NumericAverageState iterate(RamAccounting ramAccounting,
+                              MemoryManager memoryManager,
+                              NumericAverageState state,
+                              Input[] args) throws CircuitBreakingException {
+        if (state != null) {
+            BigDecimal value = returnType.implicitCast(args[0].value());
+            if (value != null) {
+                BigDecimal newValue = state.sum.value().add(value);
+                ramAccounting.addBytes(NumericType.sizeDiff(newValue, state.sum.value()));
+                ((BigDecimalValueWrapper) state.sum).setValue(newValue);
+                state.count++;
```

Comment:
What about adding `add` + `substract` to the `NumericAverageState` instead? This would move any concrete logic into impl...
```suggestion
                BigDecimal oldValue = state.sum.value();
                state.add(value);
                ramAccounting.addBytes(NumericType.sizeDiff(oldValue, state.sum.value()));
```